python/cog/server/redis_queue.py

```diff
@@ -220,16 +220,24 @@ def handle_message(self, response_queue, message, cleanup_functions):
             return_value = self.predictor.predict(**input_obj.dict())
         if isinstance(return_value, types.GeneratorType):
             output = []
+            done = False
 
-            while True:
+            while not done:
                 # we consume iterator manually to capture log
                 try:
                     with self.capture_log(self.STAGE_RUN, prediction_id), timeout(
                         seconds=self.predict_timeout, elapsed=time.time() - start_time
                     ):
                         result = next(return_value)
-                except StopIteration:
-                    break
+                except StopIteration as e:
+                    # handle returns in generator functions. Returned values, if any, are
+                    # available in StopIteration.value
+                    if e.value is None:
+                        break
+                    else:
+                        result = e.value
+                        done = True
```

Comment:
I wonder if there's a better way to structure this code.

Looking at some lines below, we actually send the final result twice because we can't tell if an iterator is finished unless we run the code between `yield` and the end of the function.

But, if there's a return value, we actually know for sure that the function is finished, so we can just send the `push_result(status="success")` and skip the `status="processing"` duplicate.

Maybe this would be better as calling a function that sends results under certain circumstances, or something.

Not a big deal, but might be worth adding as a comment.

Reply:
Yeah I was confused by the final push after the loop. What's the purpose of that?